src/operator/numpy/np_kron-inl.h

```diff
@@ -0,0 +1,261 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file np_kron-inl.h
+ * \brief Function definition of matrix numpy-compatible kron operator
+ */
+#ifndef MXNET_OPERATOR_NUMPY_NP_KRON_INL_H_
+#define MXNET_OPERATOR_NUMPY_NP_KRON_INL_H_
+
+#include <vector>
+#include "np_tensordot_op-inl.h"
+#include "../mxnet_op.h"
+
+namespace mxnet {
+namespace op {
+
+template<int ndim, int req>
+struct kron{
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(index_t i, DType* out,
+                                  const DType* a, const DType* b,
+                                  mshadow::Shape<ndim> ashape,
+                                  mshadow::Shape<ndim> bshape,
+                                  mshadow::Shape<ndim> oshape) {
+    using namespace mxnet_op;
+
+    auto k = unravel(i, oshape);
+    Shape<ndim> ia;
+    Shape<ndim> jb;
+    for(int q = 0; q < ndim; q++){
+      ia[q] = int(k[q] / bshape[q]);
+      jb[q] = k[q] % bshape[q];
+    }
+    auto idx_a = ravel(ia, ashape);
+    auto idx_b = ravel(jb, bshape);
+
+    KERNEL_ASSIGN(out[i], req, a[idx_a] * b[idx_b]);
+  }
+};
+
+template<int ndim, int req>
+struct kron_back_a{
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(index_t i, DType* agrad,
+                                  const DType* b, const DType* ograd,
+                                  mshadow::Shape<ndim> ashape,
+                                  mshadow::Shape<ndim> bshape,
+                                  mshadow::Shape<ndim> oshape) {
+    using namespace mxnet_op;
+
+    auto ia = unravel(i, ashape);
+    Shape<ndim> k;
+    DType temp_agrad = 0;
+
+    for(int idx_b = 0; idx_b < bshape.Size(); idx_b++){
+      auto jb = unravel(idx_b, bshape);
+      for(int q = 0;q < ndim; q++){
+        k[q] = ia[q]*bshape[q] + jb[q];
+      }
+      auto idx_o = ravel(k, oshape);
+      temp_agrad += b[idx_b]*ograd[idx_o];
+    }
+    KERNEL_ASSIGN(agrad[i], req, temp_agrad);
+
+  }
+};
+
+template<int ndim, int req>
+struct kron_back_b{
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(index_t i, const DType* a,
+                                  DType* bgrad, const DType* ograd,
+                                  mshadow::Shape<ndim> ashape,
+                                  mshadow::Shape<ndim> bshape,
+                                  mshadow::Shape<ndim> oshape) {
+    using namespace mxnet_op;
+
+    auto jb = unravel(i, bshape);
+    Shape<ndim> k;
+    DType temp_bgrad = 0;
+
+    for(int idx_a = 0; idx_a < ashape.Size(); idx_a++){
+      auto ia = unravel(idx_a, ashape);
+      for(int q = 0;q < ndim; q++){
+        k[q] = ia[q] * bshape[q] + jb[q];
+      }
+      auto idx_o = ravel(k, oshape);
+      temp_bgrad += a[idx_a]*ograd[idx_o];
+    }
+    KERNEL_ASSIGN(bgrad[i], req, temp_bgrad);
+  }
+};
+
+template<typename xpu>
+void KronOpForwardImpl(const OpContext& ctx,
```

Comment:
alignment:
```c++
void KronOpForwardImpl(const OpContext& ctx,
                       const TBlob& a,
                       const TBlob& b,
                       const TBlob& out,
                       const OpReqType req) {
```
Here since you're only using `req[0]` in this function I would suggest changing your function signature to look like this and call it like `KronOpForwardImpl(ctx, a, b, out, req[0]);` below.

Reply:
Fixed, thx

Reply:
seems like still not fixed on this version

Reply:
Sorry for missing the alignment, already fixed!
Thx